llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -22137,6 +22137,22 @@ static SDValue performExtendCombine(SDNode *N,
       N->getOperand(0)->getOpcode() == ISD::SETCC)
     return performSignExtendSetCCCombine(N, DCI, DAG);
 
+  // If we see (any_extend (bswap ...)) with bswap returning an i16, we know
+  // that the top half of the result register must be unused, due to the
+  // any_extend. This means that we can replace this pattern with (rev16
+  // (any_extend ...)). This saves a machine instruction compared to (lsr (rev
+  // ...)), which is what this pattern would otherwise be lowered to.
+  if (N->getOpcode() == ISD::ANY_EXTEND &&
+      N->getOperand(0).getOpcode() == ISD::BSWAP &&
+      N->getOperand(0).getValueType().isScalarInteger() &&
+      N->getOperand(0).getValueType().getFixedSizeInBits() == 16) {
```

Comment:
I think this could check `N->getOperand(0).getValueType() == MVT::i16`. Should it check that the input type is i32 too? Otherwise it would need to check i64 or i32, and add a pattern for i64 (and a test).

Reply:
Thanks for the suggestion on simplifying to `== MVT::i16`.

Reply:
By input type, I assume you mean the input type to the rev16 / output type of the old `any_extend`? As far as I can tell, this is only ever an i32. If an i64 is needed than a `zero_extend` is generated instead. 
In the absence of writing a test, do you think it would be worthwhile to write a comment explaining this? 
Regardless, using `N->getValueType(0)` instead of `EVT(MVT::i32)`, as you suggested, means the logic in the code is independent of whether it is i64 or i32, and so doesn't need a separate pattern in the code.

Reply:
I think Dave's suggestion is to add a check:

     bswap->getOperand(0).getValueType() == MVT::i32

where `bswap = N->getOperand(0)`. 

That's a very good point, it guarantees that we are only dealing with REV16 for i32 where we know the upper bits are zero. It avoids us having to worry about i64s and different patterns.

Reply:
I'm not sure I understand. `bswap`'s output type is always its input type, so `bswap->getOperand(0).getValueType()` is always i16, not i32. I just verified this by adding a check that `N->getOperand(0)->getOperand(0).getValueType() == MVT::i32`, and it made the test that checks `rev16` is generated fail i.e. we generate the old `rev` and `lsr` instead of `rev16`, because the check will evaluate to false.
We guarantee we are only dealing with REV16 for i32 because we insert an `any_extend` before the REV16, and that `any_extend` is guaranteed to be extending to i32.

Reply:
I'm worried about `anyext(bswap)` coming up for other types. These two look like they should end up with `i64 anyext` and `i128 anyext` after a little optimization: https://godbolt.org/z/G1Kc9K1cv

Reply:
OK, got it. I didn't realise that we can also get an `any_extend` from an optmised `zero_extend`.
I can, as suggested, avoid this by adding a check that the old `any_extend` output type / `rev16` input type is i32, i.e. add a check that `N.getValueType() == MVT::i32`.
But it might also be relatively simple to handle the i64 and i128 cases and still get the optimised `rev16` codegen. We could insert two `any_extend`s instead of one: one immediately before the `rev16` that extends i16 to i32, and one immediately after the `rev16` that extends i32 to whatever the value type of the old `any_extend` was (i32, i64, i128...).
Would you be happy with the second option too?

Reply:
Yeah, Adding i64 sounds like a good idea. We can just add a new tablegen pattern for REV16Xr if I understand it correctly.

Usually with the way types legalize, if you handle i32 and i64 then the other types will follow suit, as they get legalized to i32/i64 anyway. We just need to make sure we don't generate a i128 AArch64ISD::REV16 that cannot be selected.